### PR TITLE
Hw 4

### PR DESCRIPTION
**Pull request by quantum-noise**

### CODE REVIEW

File: src/http-servers/html-server.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  htmlTemplate = fs.readFileSync(filePath);
+} catch (err) {
+  console.error(err);
+}
+
+const server = http.createServer((request, response) => {
+  response.writeHead(200, {
+    'Content-Type': 'text/html',
+  });
+  response.end(htmlTemplate.replace('{message}', `Current time: ${Date.now().toString()}`));
```
Comment (boccob) [inline src/http-servers/html-server.js]: `htmlTemplate` is a Buffer without `replace` method. Did you try it yourself?

File: src/middlewares/cookies.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+export default function cookiesParser(req, res, next) {
+  req.parsedCookies = R.pipe(
+    string => string.split(';'),
+    R.map(R.pipe(
+      kv => kv.split('='),
+      ([k, v]) => [R.trim(k), v],
+    )),
+    R.fromPairs,
+  )(decodeURIComponent(req.get(cookieHeader)));
```
[the remark this answers is not in the document]
Comment (quantum-noise) [reply src/middlewares/cookies.js]: req.headers? didn't see it anywhere. i don't use 3rd party middlewares.

File: src/middlewares/query.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    R.ifElse(
+      R.propEq('length', 2),
+      R.pipe(
+        ([, query]) => query.split('&'),
+        R.map(kv => kv.split('=')),
+        R.fromPairs,
+      ),
+      () => ({}),
+    ),
+  )(decodeURIComponent(req.originalUrl));
```
[the remark this answers is not in the document]
Comment (quantum-noise) [reply src/middlewares/query.js]: req.url.query? have no ideas what is it.

File: src/http-servers/html-server.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import http from 'http';
```
[the remark this answers is not in the document]
Comment (quantum-noise) [reply src/http-servers/html-server.js]: done

File: src/app.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    res.status(401).end();
+  })
+  .get((req, res) => {
+    res.json(dao.getProducts());
+  });
+
+router
+  .route('/api/products/:id')
+  .get((req, res) => {
+    res.json(dao.getProduct(req.params.id));
```
[the remark this answers is not in the document]
Comment (quantum-noise) [reply src/app.js]: fixed in hw-5

File: src/middlewares/cookies.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (boccob) [reply src/middlewares/cookies.js]: https://nodejs.org/api/http.html#http_class_http_incomingmessage

File: src/middlewares/query.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (boccob) [reply src/middlewares/query.js]: Nevermind. You have an error with url like `?abc=%ab`. But probably it's ok.

**Pull request merged**